「Railsでテストを書く」の課題

ham-cap opened the pull request:
「Railsでテストを書く」の課題を提出です。
よろしくお願いいたします。

ham-cap committed:
ERBLintを導入し検出されたエラーをオートコレクトで修正

ham-cap committed:
プラクティスに必要のないfixturesをコメントアウトした・generatorでuser_testを生成

ham-cap committed:
ReportとUserのモデルテストを作成

ham-cap committed:
ハンズオン動画に倣ってbooksのシステムテストを作成

ham-cap committed:
reportsのシステムテストを作成

ham-cap committed:
rubocopのoffencesをオートコレクトで修正

ham-cap committed:
不要な行を削除

betachelsea commented on test/models/report_test.rb:
```diff
@@ -3,7 +3,18 @@
+    user1 = User.new(email: 'foo@example.com')
+    user2 = User.new(email: 'bar@example.com')
+    report = Report.new(user_id: user1.id, title: 'テスト', content: 'foobar')
+    assert report.editable?(user1.id)
```
[MUST] `app/models/report.rb` の実装では、idを渡すのではなく `report.editable?(user1)` のようにすることを想定されている様子なので、`user1.id` を引数として渡し、それが正常に動いている…ということは逆に問題であるといえそうです。もしかして、 `report.editable?(user1)` とやるとfalseが返ってきて困ったなどがありましたか？
`User.new` だけだと、まだDBに保存する処理が発生していないため、idは発行されません。そのため、 `user1.id` は nil が返ってくると思います。 `binding.irb` や debug gem でぜひ確認してみてください。 

`editable?` の実装をみると、IDには頼っていないのでReport.newの渡し方次第では検査できそうではありますが、保存されたデータを使ってやったほうが確実かもしれません。 fixtures を使うのもいいですね。

betachelsea commented on test/models/report_test.rb:
```diff
@@ -3,7 +3,18 @@
+    user1 = User.new(email: 'foo@example.com')
+    report = Report.new(user_id: user1.id, title: 'テスト', content: 'foobar', created_at: DateTime.new(1993, 2, 24, 12, 30, 45))
+    date = DateTime.new(1993, 2, 24, 12, 30, 45).to_date
+    assert_equal date, report.created_on
```
[nits] `to_date` した結果である、 `Date.new(1993, 2, 24)` と比較するとより分かりやすくなりそうだなと思いました :eyes:

betachelsea commented on test/models/user_test.rb:
```diff
@@ -3,7 +3,38 @@
+    user1 = User.create!(email: 'foo@example.com', password: 'password')
+    user2 = User.create!(email: 'bar@example.com', password: 'password')
+    user1.follow(user2)
+    assert user2.followed_by?(user1)
```
follow の実行前後で、`followed_by?` の結果が変わることを確認すると、より結果として間違いないものと判断できるのでオススメです。また、 1メソッドにつき1つだけ `test do 〜 end` を定義しないといけないというルールは無いので、`follow` メソッドの機能も合わせて検査した、という形にしてもOKです。

[MUST] follow関係のメソッド類について、「実行前後の変化を調べる」に着目してブラッシュアップしてみていただきたいです。

betachelsea commented on test/models/user_test.rb:
```diff
@@ -3,7 +3,38 @@
+    user = User.new(email: 'foo@example.com')
+    assert_equal 'foo@example.com', user.name_or_email
+    user.name = 'bar'
+    assert_equal 'bar', user.name_or_email
```
変化がハッキリ分かって良いですね！ 👍

betachelsea commented on test/system/books_test.rb:
```diff
@@ -5,43 +5,44 @@
+    fill_in 'メモ', with: @book.memo
+    fill_in '著者', with: @book.author
+    click_on '登録する'
+    assert_text '本が作成されました。'
```
[nits] 意図通りの本が作成されているかの確認をしておきたいですね。

betachelsea commented on test/system/books_test.rb:
```diff
@@ -5,43 +5,44 @@
     end
 
-    assert_text 'Book was successfully destroyed'
+    assert_text '本が削除されました。'
```
[nits] ちゃんとデータが消えていることを確認できるとバッチリですね。

betachelsea commented on test/system/books_test.rb:
```diff
@@ -5,43 +5,44 @@
+    click_on '編集', match: :prefer_exact
+    fill_in 'タイトル', with: @book.title
+    fill_in 'メモ', with: @book.memo
+    fill_in '著者', with: @book.author
```
[MUST] 編集するものと、編集後の指定 `@book.title` などの内容はまったく同じものである様子です。編集機能の確認なので、全く別の値を入れ、データの内容が変わっていることを確認すると間違いなさそうです。

betachelsea commented on test/system/reports_test.rb:
```diff
@@ -0,0 +1,45 @@
+    fill_in 'タイトル', with: @report.title
+    fill_in '内容', with: @report.content
+    click_on '登録する'
+    assert_text '日報が作成されました。'
```
[nits] 日報についても、`test/system/books_test.rb` と同様の内容が考慮されているとより良いですね！

betachelsea reviewed (commented):
提出ありがとうございます！コメントしました〜 📝

ham-cap committed:
report_test.rbで使用するユーザーのデータをfixturesのものに変更・created_onメソッドのテストで出力の比較対象となるオブジェクトのクラスを変更

・report_test.rbで使用するユーザーのデータはnewを使ってその場で生成していたが、DBへ保存されずidも付番されないことから、fixturesのデータを使用することとした
・created_onの出力結果をDateTimeクラスのインスタンスにto_dateを使用した結果と比較していたが、よりわかりやすさを重視するため最初からDateクラスのインスタンスとして生成することとした

ham-cap committed:
user_test.rb内で使用するユーザーのデータをfixturesのものに変更・各メソッドの実行前後の変化が分かるように修正・followは他のメソッドのテスト内で動作確認ができるため削除

ham-cap committed:
Booksのシステムテストの内容を修正

・新規作成や更新の際に使用するデータはfixturesを使用していたが、テストの開始時点でそれらのデータが登録された状態となってしまうため適切にテストできていなかった。そのため新規作成はfixturesを参照せずに文字列をそのまま使用し、更新と削除はfixturesに1件だけ用意したデータに対して実施することとした。

ham-cap committed:
テストの名称に誤りがあったため修正

ham-cap committed:
Reportsのシステムテストの内容を修正

・Booksのシステムテスト修正のコミット（d798d2abb9c4e03343ed26f5d9f0c08bc7309e81）と同様に新規作成ではfixturesを使用しないこととした

ham-cap commented:
レビューありがとうございます！
遅くなりましたが、修正してみましたので再度ご確認いただければ幸いです🙏
Booksのシステムテストの内容を修正のコミットメッセージにも書きましたが、fixturesに用意したデータの値を使って新規作成や更新をしていたことで、きちんとテストが出来ていなかったようなので`test/system/books_test.rb`と`test/system/reports_test.rb`についてはそのあたりも含めて修正しています👀

betachelsea commented on test/models/user_test.rb:
```diff
@@ -3,7 +3,34 @@
+  test '#following?' do
+    alice = users(:alice)
+    bob = users(:bob)
+    assert_not alice.following?(bob)
```
[IMO] `#unfollow` のテストの時に同一の流れがあるので、まとめてやっています、ということにしても良さそうです。 ( そうする場合は、テストの命名は変更したほうがよさそうですが :eyes: )

ham-cap replied:
たしかによく見ると同じ流れですね🤭
次回からはこういうものも注意してみてみます👀

betachelsea reviewed (approved):
修正ありがとうございます！確認OKです 👍
